Add features to the YOLO model from the latest YOLO variants

senarvi opened the pull request:
## What does this PR do?

The YOLO model has been largely refactored, so that it's easy to incorporate new features from different YOLO variants, such as different algorithms for assigning targets to anchors. It also supports defining the network structure as a PyTorch module, in addition to loading a Darknet configuration file. The implementation now supports the most important features of YOLOv3, YOLOv4, YOLOv5, YOLOv7, Scaled-YOLOv4, and YOLOX.

* Supports several new algorithms for matching targets to anchors.
* Adds support for DIoU and CIoU losses.
* By default the target for confidence of a detection that has been assigned a target is 1.0. This pull request adds a support for using the overlap between the target and the predicted box instead.
* The code is refactored from one big class into model, anchor matching, and loss function classes.
* Target class labels may be specified as a matrix of class probabilities, allowing multiple classes per object.
* Automatic padding in convolutional and max pooling layers works now in every case.
* Weight decay is applied only to convolutional layer weights.
* Supports label smoothing.
* Command line interface is now using LightningCLI.
* Network architectures can be written as PyTorch modules. YOLOv4, YOLOv5, YOLOv7, and YOLOX architectures are included.
* Supports deep supervision (auxiliary detection heads).
* Calculates MAP metric using TorchMetrics.
* Adds complete type hints for static type checking.
* Removes iou and giou metrics and losses from Bolts, as they are available from Torchvision.

Fixes #816 

## Before submitting

- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements) - issue for discussion: #816 
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review

- [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

otaj commented:
Hi @senarvi, we're currently going through a major, community driven refactor, as outlined in #839. Would you like to work with @luca-medeiros as he's the one who started working on the revision for YOLO models? Thanks a lot!

senarvi commented:
> Hi @senarvi, we're currently going through a major, community driven refactor, as outlined in #839. Would you like to work with @luca-medeiros as he's the one who started working on the revision for YOLO models? Thanks a lot!

Sounds good!

luca-medeiros commented:
@heimish-kyma

This PR brings some cool new features but it also changed a lot of stuff that you reviewed for #839.
I think we can work together to pass this PR improving YOLO while complying with the revision. Maybe even go and add v7 features as well.
I'll do a review later but seems that we can mostly improve on docs! How to effectively use the networks with the yolo class etc

senarvi commented:
In the vein of not duplicating code between other Torch codebases, I switched to using the new distance_box_iou() and complete_box_iou() functions. I found out that Torchvision also includes separate iou functions for loss computation. These are now used for the training loss, while the normal (pairwise) iou functions are used for the SimOTA cost.

The code checks whether those functions exist and only supports the losses that are included in the installed Torchvision version. Unfortunately there was a [bug](https://github.com/pytorch/vision/issues/6317) in those functions that only recently got fixed. So it's quite difficult to ensure that all losses work.

senarvi commented:
> I'll do a review later but seems that we can mostly improve on docs! How to effectively use the networks with the yolo class etc

@luca-medeiros I added a little bit to the docs, but I think an example is probably easier to understand than a lot of explanation, so I modified the CLI application so that it also shows how to use a torch network. If a Darknet configuration is not provided, it will use the YOLOv4 network. There are various other networks that can be used, but they are used in the same way. It's quite simple actually: https://github.com/groke-technologies/pytorch-lightning-bolts/blob/yolo-update/pl_bolts/models/detection/yolo/yolo_module.py#L524
Let me know if you think that something needs better explanation.

luca-medeiros commented:
@senarvi 

Sorry for the delay. I checked the changes and stuff looks awesome. 
I'm writing a review asap

luca-medeiros commented on pl_bolts/models/detection/yolo/torch_networks.py:
```diff
@@ -0,0 +1,1264 @@
+        backbone: A backbone network that returns the output from each stage.
+        width: The number of channels in the narrowest convolutional layer. The wider convolutional layers will use a
+            number of channels that is a multiple of this value.
+        depth: Repeat the bottleneck layers this many times. Can be used to make the network deeper.
```
A depth of 3 would represent which variant? Not sure if it is well documented somewhere else, but we could try to show depth-to-variant settings

senarvi replied:
> I'm not sure if we are still doing it but adding `catch_warnings` to all tests would be helpful.
> 
> An idea for bolts in general would have a README for each model folder. I think we could write the training process of the different YOLO networks there, maybe upload pretrained weights, etc.

I added `catch_warnings` to all tests and a README.

senarvi replied:
I listed the settings for the different variants.

luca-medeiros commented on pl_bolts/models/detection/yolo/torch_networks.py:
```diff
@@ -0,0 +1,1264 @@
+    Args:
+        num_classes: Number of different classes that this model predicts.
+        backbone: A backbone network that returns the output from each stage.
+        width: The number of channels in the narrowest convolutional layer. The wider convolutional layers will use a
+            number of channels that is a multiple of this value.
+        depth: Repeat the bottleneck layers this many times. Can be used to make the network deeper.
```
I guess the same works for here.

luca-medeiros commented on pl_bolts/models/detection/yolo/loss.py:
```diff
@@ -0,0 +1,302 @@
+
+        if not callable(overlap_func):
+            raise ValueError(
+                f"Unsupported overlap function '{overlap_func}'. Try upgrading Torcvision or using another IoU "
```
Small typo 
```suggestion
                f"Unsupported overlap function '{overlap_func}'. Try upgrading Torchvision or using another IoU "
```

otaj replied:
We're hitting this error in GPU CI `ValueError: Unsupported overlap function 'None'. Try upgrading Torcvision or using another IoU algorithm.`

The versions of `torch` and `torchvision` used in the CI are:
* `torch==1.11.0+cu113`
* `torchvision==0.12.0+cu113`

senarvi replied:
Now the tests use giou loss, which is available in torchvision 0.12. I also improved the error message.

luca-medeiros reviewed (approved):
LGTM! Good work, couldn't catch any major issue.
I'm not sure if we are still doing it but adding `catch_warnings` to all tests would be helpful.

An idea for bolts in general would have a README for each model folder.  I think we could write the training process of the different YOLO networks there, maybe upload pretrained weights, etc.

otaj reviewed (commented):
Oh my, that's a large PR to review :joy:

Overall, I like the code, though I have to say I haven't combed through it with a very fine comb :smile: One major thing at the moment is if you can add `catch_warnings` fixture to all the tests that are in this PR, so that we are sure no warnings are issued.

Also, I see couple failing tests, and I don't have much of an idea of what is causing them, can you please merge master to see if it helps?

otaj commented on pl_bolts/models/detection/yolo/utils.py:
```diff
@@ -0,0 +1,155 @@
+
+# PyTorch 1.10 introduced the argument "indexing" and deprecated calling without the argument. Since we call it inside
+# a "@torch.jit.script" function, it's difficult to make this decision at call time.
+if version.parse(torch.__version__) >= version.parse("1.10.0"):
```
This looks like it's throwing an error in one of the tests, please use `_compare_version` from `pl_bolts.utils`

senarvi replied:
This now uses _TORCH_MESHGRID_REQUIRES_INDEXING, as it was fixed in the other pull request.

luca-medeiros reviewed (approved):
Awesome!

senarvi commented on pl_bolts/models/detection/yolo/yolo_module.py:
```diff
@@ -1,613 +1,586 @@
 import torch.nn as nn
 from pytorch_lightning import LightningModule
-from pytorch_lightning.utilities import rank_zero_info
+from pytorch_lightning.cli import LightningCLI
```
I changed this.

senarvi commented:
> Overall, I like the code, though I have to say I haven't combed through it with a very fine comb 😄 One major thing at the moment is if you can add `catch_warnings` fixture to all the tests that are in this PR, so that we are sure no warnings are issued.
> 
> Also, I see couple failing tests, and I don't have much of an idea of what is causing them, can you please merge master to see if it helps?

The tests are passing now and `catch_warnings` have been added. There are only some build failures related to other models.

senarvi commented:
I noticed that the iou and giou losses are also included in Bolts: https://github.com/Lightning-AI/lightning-bolts/blob/master/pl_bolts/losses/object_detection.py

But the plan was to reduce code duplication between different PyTorch projects, so I would remove them from Bolts and use the Torchvision functions.

otaj commented:
> I noticed that the iou and giou losses are also included in Bolts: https://github.com/Lightning-AI/lightning-bolts/blob/master/pl_bolts/losses/object_detection.py
> 
> But the plan was to reduce code duplication between different PyTorch projects, so I would remove them from Bolts and use the Torchvision functions.

Yes, please!

senarvi commented:
> > I noticed that the iou and giou losses are also included in Bolts: https://github.com/Lightning-AI/lightning-bolts/blob/master/pl_bolts/losses/object_detection.py
> > But the plan was to reduce code duplication between different PyTorch projects, so I would remove them from Bolts and use the Torchvision functions.
> 
> Yes, please!

Should I remove them in this pull request?

otaj commented:
> > > I noticed that the iou and giou losses are also included in Bolts: https://github.com/Lightning-AI/lightning-bolts/blob/master/pl_bolts/losses/object_detection.py
> > > But the plan was to reduce code duplication between different PyTorch projects, so I would remove them from Bolts and use the Torchvision functions.
> > 
> > 
> > Yes, please!
> 
> Should I remove them in this pull request?

Yes, @senarvi, sorry for not being clearer. It would be great if you could manage the removing in this PR. I could help, but it seems I don't have write access to the PR

senarvi commented:
> Yes, @senarvi, sorry for not being clearer. It would be great if you could manage the removing in this PR. I could help, but it seems I don't have write access to the PR

No problem. I removed them. And if there's a way to give you write access, I'm happy to do so.

otaj reviewed (approved):
Hi, @senarvi, I was quite busy lately. Overall it looks good, could you please merge master and then we should be good to go

senarvi commented:
> could you please merge master and then we should be good to go

@otaj Done.

otaj commented:
Hi, @senarvi! Thank you, there are currently two issue.

1. `pre-commit` is complaining and it isn't capable to automatically send fixes for your PR, because it has the same limitations as I do - it cannot contribute to org repository. That has a very simple fix: run `pre-commit run -a` on the repository and commit result. More robust for the future is to have `pre-commit` installed so it runs on every commit for you. That would be done by `pre-commit install`
2. The other issue is a bit tougher to resolve but shouldn't be too hard. When using DDP, you can't use `lambda` functions, because those are unable to be passed through pickling process. If you can get rid of that `lambda` function in `_get_iou_and_loss_functions` it should pass now.

senarvi commented:
> 1. `pre-commit` is complaining and it isn't capable to automatically send fixes for your PR, because it has the same limitations as I do - it cannot contribute to org repository. That has a very simple fix: run `pre-commit run -a` on the repository and commit result. More robust for the future is to have `pre-commit` installed so it runs on every commit for you. That would be done by `pre-commit install`
> 2. The other issue is a bit tougher to resolve but shouldn't be too hard. When using DDP, you can't use `lambda` functions, because those are unable to be passed through pickling process. If you can get rid of that `lambda` function in `_get_iou_and_loss_functions` it should pass now.

@otaj great to know about `pre-commit`! Those things should be fixed now.

senarvi commented:
As @luca-medeiros said, I think it would be interesting to look into new features from YOLOv7. But to keep things in control, let's first merge this old pull request. Do we still need one more review or are we good to go @otaj ?

otaj commented:
Hi @senarvi,

you already have my approving review for quite some time, however, since I've been laid off from the company in November, my review lost all its powers as I suddenly don't have any privileges with regards to the repo. For anything regarding this repo, contact @Borda

senarvi commented:
> Hi @senarvi,
> 
> you already have my approving review for quite some time, however, since I've been laid off from the company in November, my review lost all its powers as I suddenly don't have any privileges with regards to the repo. For anything regarding this repo, contact @Borda

Got it. Thanks for the help @otaj and best of luck!

@Borda how should we proceed with merging this PR?

senarvi commented:
I added a YOLOv7 architecture with deep supervision. It was actually quite straightforward to add. Just had to separate the forward pass and loss calculation a bit better in the `DetectionLayer` class. The tests are failing because of this issue: https://github.com/Lightning-AI/lightning-bolts/issues/962

Borda commented:
@senarvi we have fixed master a few munites ago, so shall be 🟢 could you pls merge it here 🐿️

senarvi commented:
> @senarvi we have fixed master a few munites ago, so shall be 🟢 could you pls merge it here 🐿️

Great. Seems like the tests are passing now.

Borda commented:
> > @senarvi we have fixed master a few munites ago, so shall be 🟢 could you pls merge it here 🐿️
> 
> Great. Seems like the tests are passing now.

perfect, it just sai about some conflicts, but I can't see them, mind check it?
Then I set ti on auto-merge so shall by landing any time soo 🎉

senarvi commented:
> > > @senarvi we have fixed master a few munites ago, so shall be 🟢 could you pls merge it here 🐿️
> > 
> > 
> > Great. Seems like the tests are passing now.
> 
> perfect, it just sai about some conflicts, but I can't see them, mind check it? Then I set ti on auto-merge so shall by landing any time soo 🎉

@Borda managed to fix it finally. ruff formatted the code in a way that caused mypy to fail. Formatting in a way that mypy accepts caused ruff to fail. In the end I added a "type: ignore" comment. It wasn't merged automatically, but all the tests have passed.

Borda commented on src/pl_bolts/models/detection/yolo/README.md:
```diff
@@ -0,0 +1,59 @@
+# YOLO
+
+The YOLO model has evolved quite a bit, since the original publication in 2016. The original source code was written in C, using a framework called [Darknet](https://github.com/pjreddie/darknet). The final revision by the original author was called YOLOv3 and described in an [arXiv paper](https://arxiv.org/abs/1804.02767). Later various other authors have written implementations that improve diff ...
```
cc: @lantiga

lantiga reviewed (approved):
Very cool contribution, thank you @senarvi
